### ISSUE TITLE
A keyword named variable crashes the compiler

### ISSUE THREAD

**Issue by RandomShaper** [microsoft/DirectXShaderCompiler#4310]
During my usage of DXC, I've came across a shader that happens to call a variable `unorm`. That triggers a failed assertion abort with the message, `Not a simple-type-specifier token!`.

I'm not sure now if a keyword is a valid identifier; MSDN docs won't tell. Regardless, I guess this should cause a compiler error instead of an internal error.

In case it's needed, here's the full code of the shader. The offending line is 153.
```
struct Light
{
    uint type;
    float energy;
    float radius;
    float attenuation;
    float3 color;
    float cos_spot_angle;
    float3 position;
    float inv_spot_attenuation;
    float3 direction;
    uint has_shadow;
};

struct CellData
{
    uint position;
    uint albedo;
    uint emission;
    uint normal;
};

struct CellChildren
{
    uint children[8];
};

cbuffer Lights
{
    Light lights_data[32] : packoffset(c0);
};

RWByteAddressBuffer cell_data;
RWByteAddressBuffer outputs;
RWByteAddressBuffer cell_children;
cbuffer Params
{
    int3 params_limits : packoffset(c0);
    uint params_stack_size : packoffset(c0.w);
    float params_emission_scale : packoffset(c1);
    float params_propagation : packoffset(c1.y);
    float params_dynamic_range : packoffset(c1.z);
    uint params_light_count : packoffset(c1.w);
    uint params_cell_offset : packoffset(c2);
    uint params_cell_count : packoffset(c2.y);
    float params_aniso_strength : packoffset(c2.z);
    uint params_pad : packoffset(c2.w);
};

Texture3D<float4> texture_sdf;
SamplerState texture_sampler;

static uint3 gl_GlobalInvocationID;
struct SPIRV_Cross_Input
{
    uint3 gl_GlobalInvocationID : SV_DispatchThreadID;
};

float mod(float x, float y)
{
    return x - y * floor(x / y);
}

float2 mod(float2 x, float2 y)
{
    return x - y * floor(x / y);
}

float3 mod(float3 x, float3 y)
{
    return x - y * floor(x / y);
}

float4 mod(float4 x, float4 y)
{
    return x - y * floor(x / y);
}

uint spvPackUnorm4x8(float4 value)
{
    uint4 Packed = uint4(round(saturate(value) * 255.0));
    return Packed.x | (Packed.y << 8) | (Packed.z << 16) | (Packed.w << 24);
}

float4 spvUnpackUnorm4x8(uint value)
{
    uint4 Packed = uint4(value & 0xff, (value >> 8) & 0xff, (value >> 16) & 0xff, value >> 24);
    return float4(Packed) / 255.0;
}

uint spvPackSnorm4x8(float4 value)
{
    int4 Packed = int4(round(clamp(value, -1.0, 1.0) * 127.0)) & 0xff;
    return uint(Packed.x | (Packed.y << 8) | (Packed.z << 16) | (Packed.w << 24));
}

float4 spvUnpackSnorm4x8(uint value)
{
    int SignedValue = int(value);
    int4 Packed = int4(SignedValue << 24, SignedValue << 16, SignedValue << 8, SignedValue) >> 24;
    return clamp(float4(Packed) / 127.0, -1.0, 1.0);
}

float get_omni_attenuation(float _distance, float inv_range, float decay)
{
    float nd = _distance * inv_range;
    nd *= nd;
    nd *= nd;
    nd = max(1.0f - nd, 0.0f);
    nd *= nd;
    return nd * pow(max(_distance, 9.9999997473787516355514526367188e-05f), -decay);
}

bool compute_light_vector(uint light, float3 pos, inout float attenuation, inout float3 light_pos)
{
    if (lights_data[light].type == 0u)
    {
        light_pos = pos - (lights_data[light].direction * length(float3(params_limits)));
        attenuation = 1.0f;
    }
    else
    {
        light_pos = lights_data[light].position;
        float _distance = length(pos - light_pos);
        if (_distance >= lights_data[light].radius)
        {
            return false;
        }
        float param = _distance;
        float param_1 = 1.0f / lights_data[light].radius;
        float param_2 = lights_data[light].attenuation;
        attenuation = get_omni_attenuation(param, param_1, param_2);
        if (lights_data[light].type == 2u)
        {
            float3 rel = normalize(pos - light_pos);
            float cos_spot_angle = lights_data[light].cos_spot_angle;
            float cos_angle = dot(rel, lights_data[light].direction);
            if (cos_angle < cos_spot_angle)
            {
                return false;
            }
            float scos = max(cos_angle, cos_spot_angle);
            float spot_rim = max(9.9999997473787516355514526367188e-05f, (1.0f - scos) / (1.0f - cos_spot_angle));
            attenuation *= (1.0f - pow(spot_rim, lights_data[light].inv_spot_attenuation));
        }
    }
    return true;
}

float get_normal_advance(float3 p_normal)
{
    float3 normal = p_normal;
    float3 unorm = abs(normal);
    bool _271 = unorm.x >= unorm.y;
    bool _279;
    if (_271)
    {
        _279 = unorm.x >= unorm.z;
    }
    else
    {
        _279 = _271;
    }
    if (_279)
    {
        unorm = (normal.x > 0.0f) ? float3(1.0f, 0.0f, 0.0f) : float3(-1.0f, 0.0f, 0.0f);
    }
    else
    {
        bool _294 = unorm.y > unorm.x;
        bool _302;
        if (_294)
        {
            _302 = unorm.y >= unorm.z;
        }
        else
        {
            _302 = _294;
        }
        if (_302)
        {
            unorm = (normal.y > 0.0f) ? float3(0.0f, 1.0f, 0.0f) : float3(0.0f, -1.0f, 0.0f);
        }
        else
        {
            bool _316 = unorm.z > unorm.x;
            bool _324;
            if (_316)
            {
                _324 = unorm.z > unorm.y;
            }
            else
            {
                _324 = _316;
            }
            if (_324)
            {
                unorm = (normal.z > 0.0f) ? float3(0.0f, 0.0f, 1.0f) : float3(0.0f, 0.0f, -1.0f);
            }
            else
            {
                unorm = float3(1.0f, 0.0f, 0.0f);
            }
        }
    }
    return 1.0f / dot(normal, unorm);
}

void clip_segment(float4 plane, float3 begin, inout float3 end)
{
    float3 segment = begin - end;
    float den = dot(plane.xyz, segment);
    if (den < 9.9999997473787516355514526367188e-05f)
    {
        return;
    }
    float dist = (dot(plane.xyz, begin) - plane.w) / den;
    if ((dist < 9.9999997473787516355514526367188e-05f) || (dist > 1.00010001659393310546875f))
    {
        return;
    }
    end = begin + (segment * (-dist));
}

float raymarch(inout float _distance, float distance_adv, inout float3 from, float3 direction)
{
    float3 cell_size = 1.0f.xxx / float3(params_limits);
    float occlusion = 1.0f;
    while (_distance > 0.5f)
    {
        float advance = (texture_sdf.SampleLevel(texture_sampler, from * cell_size, 0.0f).x * 255.0f) - 1.0f;
        if (advance < 0.0f)
        {
            occlusion = 0.0f;
            break;
        }
        occlusion = min(advance, occlusion);
        advance = max(distance_adv, advance - mod(advance, distance_adv));
        from += (direction * advance);
        _distance -= advance;
    }
    return occlusion;
}

bool compute_light_at_pos(uint index, float3 pos, float3 normal, inout float3 light, inout float3 light_dir)
{
    uint param = index;
    float3 param_1 = pos;
    float param_2;
    float3 param_3;
    bool _388 = compute_light_vector(param, param_1, param_2, param_3);
    float attenuation = param_2;
    float3 light_pos = param_3;
    if (!_388)
    {
        return false;
    }
    light_dir = normalize(pos - light_pos);
    bool _401 = attenuation < 0.00999999977648258209228515625f;
    bool _416;
    if (!_401)
    {
        bool _408 = length(normal) > 0.20000000298023223876953125f;
        bool _415;
        if (_408)
        {
            _415 = dot(normal, light_dir) >= 0.0f;
        }
        else
        {
            _415 = _408;
        }
        _416 = _415;
    }
    else
    {
        _416 = _401;
    }
    if (_416)
    {
        return false;
    }
    if (lights_data[index].has_shadow != 0u)
    {
        float3 param_4 = light_dir;
        float distance_adv = get_normal_advance(param_4);
        float3 to = pos;
        if (length(normal) > 0.20000000298023223876953125f)
        {
            to += ((normal * distance_adv) * 0.5099999904632568359375f);
        }
        else
        {
            to -= (sign(light_dir) * 0.449999988079071044921875f);
        }
        float4 _459 = float4(1.0f, 0.0f, 0.0f, float(params_limits.x - 1));
        bool4 _464 = (light_dir.x < 0.0f).xxxx;
        float4 param_5 = float4(_464.x ? _459.x : float4(-1.0f, 0.0f, 0.0f, 0.0f).x, _464.y ? _459.y : float4(-1.0f, 0.0f, 0.0f, 0.0f).y, _464.z ? _459.z : float4(-1.0f, 0.0f, 0.0f, 0.0f).z, _464.w ? _459.w : float4(-1.0f, 0.0f, 0.0f, 0.0f).w);
        float3 param_6 = to;
        float3 param_7 = light_pos;
        clip_segment(param_5, param_6, param_7);
        light_pos = param_7;
        float4 _478 = float4(0.0f, 1.0f, 0.0f, float(params_limits.y - 1));
        bool4 _482 = (light_dir.y < 0.0f).xxxx;
        float4 param_8 = float4(_482.x ? _478.x : float4(0.0f, -1.0f, 0.0f, 0.0f).x, _482.y ? _478.y : float4(0.0f, -1.0f, 0.0f, 0.0f).y, _482.z ? _478.z : float4(0.0f, -1.0f, 0.0f, 0.0f).z, _482.w ? _478.w : float4(0.0f, -1.0f, 0.0f, 0.0f).w);
        float3 param_9 = to;
        float3 param_10 = light_pos;
        clip_segment(param_8, param_9, param_10);
        light_pos = param_10;
        float4 _496 = float4(0.0f, 0.0f, 1.0f, float(params_limits.z - 1));
        bool4 _500 = (light_dir.z < 0.0f).xxxx;
        float4 param_11 = float4(_500.x ? _496.x : float4(0.0f, 0.0f, -1.0f, 0.0f).x, _500.y ? _496.y : float4(0.0f, 0.0f, -1.0f, 0.0f).y, _500.z ? _496.z : float4(0.0f, 0.0f, -1.0f, 0.0f).z, _500.w ? _496.w : float4(0.0f, 0.0f, -1.0f, 0.0f).w);
        float3 param_12 = to;
        float3 param_13 = light_pos;
        clip_segment(param_11, param_12, param_13);
        light_pos = param_13;
        float _distance = length(to - light_pos);
        if (_distance < 0.100000001490116119384765625f)
        {
            return false;
        }
        _distance += (distance_adv - mod(_distance, distance_adv));
        light_pos = to - (light_dir * _distance);
        float param_14 = _distance;
        float param_15 = distance_adv;
        float3 param_16 = light_pos;
        float3 param_17 = light_dir;
        float _541 = raymarch(param_14, param_15, param_16, param_17);
        float occlusion = _541;
        if (occlusion == 0.0f)
        {
            return false;
        }
        attenuation *= occlusion;
    }
    light = (lights_data[index].color * attenuation) * lights_data[index].energy;
    return true;
}

void comp_main()
{
    uint cell_index = gl_GlobalInvocationID.x;
    if (cell_index >= params_cell_count)
    {
        return;
    }
    cell_index += params_cell_offset;
    uint3 posu = uint3(cell_data.Load<uint>(cell_index * 16 + 0) & 2047u, (cell_data.Load<uint>(cell_index * 16 + 0) >> uint(11)) & 1023u, cell_data.Load<uint>(cell_index * 16 + 0) >> uint(21));
    float4 albedo = spvUnpackUnorm4x8(cell_data.Load<uint>(cell_index * 16 + 4));
    float3 pos = float3(posu) + 0.5f.xxx;
    float3 emission = float3(uint3(cell_data.Load<uint>(cell_index * 16 + 8) & 511u, (cell_data.Load<uint>(cell_index * 16 + 8) >> uint(9)) & 511u, (cell_data.Load<uint>(cell_index * 16 + 8) >> uint(18)) & 511u)) * pow(2.0f, (float(cell_data.Load<uint>(cell_index * 16 + 8) >> uint(27)) - 15.0f) - 9.0f);
    float3 normal = spvUnpackSnorm4x8(cell_data.Load<uint>(cell_index * 16 + 12)).xyz;
    float3 accum = 0.0f.xxx;
    float3 light;
    float3 light_dir;
    for (uint i = 0u; i < params_light_count; i++)
    {
        uint param = i;
        float3 param_1 = pos;
        float3 param_2 = normal;
        float3 param_3 = light;
        float3 param_4 = light_dir;
        bool _679 = compute_light_at_pos(param, param_1, param_2, param_3, param_4);
        light = param_3;
        light_dir = param_4;
        if (!_679)
        {
            continue;
        }
        light *= albedo.xyz;
        if (length(normal) > 0.20000000298023223876953125f)
        {
            accum += (light * max(0.0f, dot(normal, -light_dir)));
        }
        else
        {
            accum += light;
        }
    }
    outputs.Store<float4>(cell_index * 16 + 0, float4(accum + emission, 0.0f));
}

[numthreads(64, 1, 1)]
void main(SPIRV_Cross_Input stage_input)
{
    gl_GlobalInvocationID = stage_input.gl_GlobalInvocationID;
    comp_main();
}
```
